Transfers: metadata for collocation on tape #6398

voetberg opened the pull request:
Constructs a plugin class that adds a field to the transfer metadata, utilized by fts3 for placing like with like when storing on tape

Plugin is a subclass of `PolicyPackageAlgorithm`, so any experiment-specific plugin someone writes needs to be registered as a policy under the name "fts3_plugins" to be picked up during the initialization of the fts3 transfer tool. 

Additionally, to use a specifically plugin, it needs to be included in the rucio.config as the following: 
```
[transfers] 
plugins=<A list of the names of plugins being used, as defined during the plugin registration>
```

Included pre-build is a priority field based on activity, which produces an append to the transfer job submission in the form: 
```
{"archive_metadata": {"scheduling_hints": {"priority": "20"}}
```
Where priority is assigned based on activity, defined by a table set in the config. Default priority is 20. 
```

[transfers]
plugins=activity

[tape priority]
<activity_1>=40 
<activity_2>=100
```

rcarpa commented on lib/rucio/transfertool/fts3.py:
```diff
@@ -797,6 +798,9 @@ def __init__(self, external_host, oidc_account=None, oidc_support: bool = False,
         self.default_lifetime = default_lifetime
         self.archive_timeout_override = archive_timeout_override
 
+        plugins = config_get("transfers", "plugins", False, "[]").strip("[]").split(",")
```
You probably want to use config_get_list here

voetberg replied:
Didn't know that existed! Useful, thanks.

rcarpa commented on lib/rucio/transfertool/fts3.py:
```diff
@@ -936,6 +940,21 @@ def _file_from_transfer(self, transfer, job_params):
+            # Update the t_file
+            # Support up to 2 levels of further updates
+
+            for key in plugin_hints.keys():
```
directly iterate keys and values using items().
Ideally, implement a `deep_merge_dicts` function somewhere in rucio/coomon/utils and call it here. It will make the code easier to read rather than try to understand what exactly is happening here.

rcarpa commented on lib/rucio/transfertool/fts3_plugins.py:
```diff
@@ -0,0 +1,121 @@
+class FTS3Plugin(PolicyPackageAlgorithms):
+    """
+    Generic plugin for fts3
+    Default plugin, appends an empty dict to the job params
```
I don't see any job params being updated. You meant file params?

rcarpa commented on lib/rucio/transfertool/fts3.py:
```diff
@@ -936,6 +940,21 @@ def _file_from_transfer(self, transfer, job_params):
                 t_file['scitag'] = self.scitags_exp_id << 6 | activity_id
 
+        for plugin in self.plugins:
+            plugin_hints = PolicyPackageAlgorithms._get_one_algorithm(
```
Why are we calling a private method here ? Necessary work must be done to make it public if it should be called here.

voetberg replied:
Yep, totally right there. I made some changes to avoid this.

rcarpa commented on lib/rucio/transfertool/fts3_plugins.py:
```diff
@@ -0,0 +1,121 @@
+        Returns:
+            dict: {'prority': 'Activity Transfer Prority'}
+        """
+        prority = config_get(
```
If we start calling config_get functions for each transfer, the database session must be passed through. This will require a refactoring of the whole transfertool interface.

rcarpa commented on lib/rucio/transfertool/fts3_plugins.py:
```diff
@@ -0,0 +1,121 @@
+    def __init__(self):
+        super().__init__()
+
+        self.default_prority = config_get(
```
Isn't it an int?

rcarpa commented on tests/test_fts3_transfer_tools.py:
```diff
@@ -0,0 +1,397 @@
+
+    activity = transfer_path[0].rws.activity
+    _make_config_section()
+    config_set(section="tape_prority", option=activity, value="100")
```
Setting config values in tests like that can impact all other tests running in parallel. Check file_config_mock and core_config_mock fixtures.

voetberg replied:
Ah, didn't know about those. Changing it now.

rcarpa commented on tests/test_fts3_transfer_tools.py:
```diff
@@ -0,0 +1,397 @@
+
+    # Mock Transfer Tool
+    fts3_tool = FTS3Transfertool(TEST_FTS_HOST)
+    assert "FTS3SchedulingHints" in fts3_tool.plugins
```
I, personally, am a  very strong opponent of tests which verify the status of internal data structures. If we rename self.plugins to self.metadata_plugins later, we have to also update tests. Any simple refactoring will require updating the tests. This way, if you introduce bugs, you end also updating the test to fit the new bugs. Making the test useless. I advise to update some tests in `test_conveyor.py` by setting the correct configuration values using file_config_mock and core_config_mock fixtures. Following by checking that FTS (not rucio!) does what it's intended to do.

voetberg replied:
Honestly it's just how I was taught to do tests and sort of did it uncritically, but your logic makes sense. They also end up being redundant, so, removed. I was struggling with figuring out how to verify fts is correctly using the metadata, but I will check out `test_conveyor`, thanks!

voetberg replied:
So I did start adding conveyor tests, but I consistently get an error `root:fts3.py:1015 Could not submit transfer to https://fts:8446 - HTTPSConnectionPool(host='fts', port=8446): Max retries exceeded with url: /jobs (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f0e9c2d36d0>: Failed to establish a new connection: [Errno -2] Name or service not known'))`
Is there is a known work-around or known way to properly connect to that fts server that I'm missing?

rcarpa replied:
How are you running the tests ? You need to run them from within the docker-compose env: https://rucio.github.io/documentation/operator/setting_up_demo/#using-the-environment-including-storage 
(also see the previous section on how to run the tests). If you want to use it from a GUI and have a paid intellij licence, you can use the quite neat integration (only tested on linux): https://rucio.github.io/documentation/developer/setting_up_intellij_dev_env; or (a bit more complex to set-up, but free) vscode alternative: https://rucio.github.io/documentation/developer/setting_up_vscode_dev_env

voetberg replied:
Oh, yeah I've been using the vscode integration, no worries. Never mind though, looks like I was just mis-reading some of the error messages.

voetberg committed:
Transfers: Plugins for interacting with fts3  field #6398

bari12 commented on lib/rucio/transfertool/fts3_plugins.py:
```diff
@@ -0,0 +1,152 @@
+from rucio.common.exception import InvalidRequest
+
+
+class FTS3MetadataPlugin(PolicyPackageAlgorithms):
```
Let's rename this to `FTS3TapeMetadataPlugin`, I think this is more accurate, since it is not meant for all kinds of metadata.

bari12 commented on lib/rucio/transfertool/fts3.py:
```diff
@@ -797,6 +799,9 @@ def __init__(self, external_host, oidc_account=None, oidc_support: bool = False,
         self.archive_timeout_override = archive_timeout_override
 
+        plugins = config_get_list("transfers", "plugins", False, "[]")
+        self.plugins = [FTS3MetadataPlugin(plugin.strip(" ")) for plugin in plugins]
```
I wonder if we would ever have the case that this would be multiple different plugins.

voetberg replied:
`fts3tape_metadata_plugins` would probably be better for future proofing (and legibility)

bari12 replied:
What I meant here is: Would it ever happen that a community loads multiple plugins? I was operating under the assumption it would be 0 or 1.

bari12 commented on lib/rucio/transfertool/fts3_plugins.py:
```diff
@@ -0,0 +1,152 @@
+    Plugins are registered during initialization and called during a transfer with FTS3
+    """
+
+    ALGORITHM_NAME = "fts3_plugins"
```
Let's also go for `fts3_tape_metadata_plugins` here

bari12 commented on lib/rucio/transfertool/fts3_plugins.py:
```diff
@@ -0,0 +1,152 @@
+    def _default(self, *hints: dict) -> dict:
+        return {}
+
+    def _cms_collocation(self, *hints: dict) -> None:
```
Remove, I don't see why this would need to be mentioned in the abstract class. Should only be in the CMS specific implementation.

bari12 commented on lib/rucio/transfertool/fts3_plugins.py:
```diff
@@ -0,0 +1,152 @@
+
+    def _init_activity_hints(self) -> None:
+        try:
+            self.prority_table = dict(config_get_items("tape_priority"))
```
Maybe also some more descriptive name here

bari12 reviewed (changes requested):
This looks very good and is almost gtg 👍 
I added some very minor comments.
@rdimaio can you please have a look if there is anything to change in respect to the typing.

bari12 commented:
Oh and please rebase :-)

rdimaio commented on lib/rucio/common/utils.py:
```diff
@@ -1944,3 +1944,16 @@ def _wrapper(*args, **kwargs):
+
+
+def deep_merge_dict(source: dict, destination: dict) -> dict:
+    """Merge two dictionaries together recurively"""
```
typo here, "recurively" -> "recursively"

rdimaio commented on lib/rucio/transfertool/fts3.py:
```diff
@@ -295,7 +297,7 @@ def _pick_fts_checksum(
 
 
-def build_job_params(transfer_path, bring_online, default_lifetime, archive_timeout_override, max_time_in_queue, logger):
+def build_job_params(transfer_path: list, bring_online: Optional[int], default_lifetime: Optional[int], archive_timeout_override: Optional[int], max_time_in_queue: Optional[dict], logger: Callable) -> dict[str, Any]:
```
`transfer_path: list[DirectTransfer]` might be more accurate (I'm assuming that's the right type based on these two instances: https://github.com/rucio/rucio/blob/48de26679d1fb405745e1cc02963a9faf520c5bb/lib/rucio/transfertool/fts3.py#L247 and https://github.com/rucio/rucio/blob/48de26679d1fb405745e1cc02963a9faf520c5bb/lib/rucio/core/transfer.py#L281)

voetberg replied:
Shouldn't that be `typing.List[DirectTransfer]`? (I was just reviewing the typing guide and noticed....)

rdimaio replied:
I assume the typing guide was written before Python 3.9, when they [added support for type hinting using standard collections](https://docs.python.org/3/whatsnew/3.9.html#type-hinting-generics-in-standard-collections); @bari12 could you confirm?

(source: https://stackoverflow.com/questions/39458193/using-list-tuple-etc-from-typing-vs-directly-referring-type-as-list-tuple-etc)

dchristidis replied:
Yes, the guide is unfortunately out of date. See also #6276.

bari12 replied:
Correct

rdimaio commented on lib/rucio/transfertool/fts3.py:
```diff
@@ -388,8 +390,8 @@ def build_job_params(transfer_path, bring_online, default_lifetime, archive_time
 
-def bulk_group_transfers(transfer_paths, policy='rule', group_bulk=200, source_strategy=None, max_time_in_queue=None,
-                         logger=logging.log, archive_timeout_override=None, bring_online=None, default_lifetime=None):
+def bulk_group_transfers(transfer_paths: list, policy: str = 'rule', group_bulk: int = 200, source_strategy: Optional[str] = None, max_time_in_queue: Optional[dict] = None,
```
`transfer_paths: list[list[DirectTransfer]]` might be more accurate (I'm assuming that's the right type based on these two instances: https://github.com/rucio/rucio/blob/48de26679d1fb405745e1cc02963a9faf520c5bb/lib/rucio/transfertool/fts3.py#L247 and https://github.com/rucio/rucio/blob/48de26679d1fb405745e1cc02963a9faf520c5bb/lib/rucio/core/transfer.py#L281)

voetberg replied:
That seems to be right, according to my ide at least

rdimaio commented on lib/rucio/transfertool/fts3.py:
```diff
@@ -295,7 +297,7 @@ def _pick_fts_checksum(
 
 
-def build_job_params(transfer_path, bring_online, default_lifetime, archive_timeout_override, max_time_in_queue, logger):
+def build_job_params(transfer_path: list, bring_online: Optional[int], default_lifetime: Optional[int], archive_timeout_override: Optional[int], max_time_in_queue: Optional[dict], logger: Callable) -> dict[str, Any]:
```
For the `Optional` arguments, it might be worth adding `None` as default value (unless it's expected that this function will always have those parameters passed in, even when they're `None`).

rdimaio commented on lib/rucio/transfertool/fts3.py:
```diff
@@ -388,8 +390,8 @@ def build_job_params(transfer_path, bring_online, default_lifetime, archive_time
-def bulk_group_transfers(transfer_paths, policy='rule', group_bulk=200, source_strategy=None, max_time_in_queue=None,
-                         logger=logging.log, archive_timeout_override=None, bring_online=None, default_lifetime=None):
+def bulk_group_transfers(transfer_paths: list, policy: str = 'rule', group_bulk: int = 200, source_strategy: Optional[str] = None, max_time_in_queue: Optional[dict] = None,
+                         logger: Callable = logging.log, archive_timeout_override: Optional[int] = None, bring_online: Optional[int] = None, default_lifetime: Optional[int] = None):
```
missing return type `list[dict[str, any]]`

As a consideration - it might be worth making `{'transfers': ..., 'job_params': ...}` into a custom `TypedDict` for typing purposes (but I wouldn't necessarily do this if this type of dict is not being used anywhere other than here)

rdimaio commented on lib/rucio/transfertool/fts3.py:
```diff
@@ -509,10 +511,10 @@ def __str__(self):
                f'{self._file_metadata["src_rse"]} --({self._file_metadata["request_id"]})-> {self._file_metadata["dst_rse"]}'
 
-    def initialize(self, session, logger=logging.log):
+    def initialize(self, session: "Session", logger: Callable = logging.log):
```
please add `-> None` for the return type

rdimaio replied:
Side-note: you can use the [`@abstractmethod` decorator](https://docs.python.org/3/library/abc.html#abc.abstractmethod) to mark this method as abstract, in order to enforce its implementation in child classes

rdimaio commented on lib/rucio/transfertool/fts3.py:
```diff
@@ -509,10 +511,10 @@ def __str__(self):
         raise NotImplementedError(f"{self.__class__.__name__} is abstract and shouldn't be used directly")
 
-    def get_monitor_msg_fields(self, session, logger=logging.log):
+    def get_monitor_msg_fields(self, session: "Session", logger: Callable = logging.log):
```
return type: `dict[str, Any]`

rdimaio commented on lib/rucio/transfertool/fts3.py:
```diff
@@ -531,7 +533,7 @@ def get_monitor_msg_fields(self, session, logger=logging.log):
         return '%s/fts3/ftsmon/#/job/%s' % (self._fts_address.replace('8446', '8449'), self._transfer_id)
 
-    def _find_attribute_updates(self, request, new_state, reason, overwrite_corrupted_files):
+    def _find_attribute_updates(self, request: dict, new_state: RequestState, reason: str, overwrite_corrupted_files: Optional[bool]):
```
I would change `overwrite_corrupted_files: Optional[bool]` to `overwrite_corrupted_files: bool = False`

return type: `Optional[dict[str, Any]])`

rdimaio commented on lib/rucio/transfertool/fts3.py:
```diff
@@ -541,7 +543,7 @@ def _find_attribute_updates(self, request, new_state, reason, overwrite_corrupte
         return attributes
 
-    def _find_used_source_rse(self, session, logger):
+    def _find_used_source_rse(self, session: "Session", logger: Callable):
```
return type: `tuple[Optional[str], Optional[str]]`

rdimaio commented on lib/rucio/transfertool/fts3.py:
```diff
@@ -559,7 +561,7 @@ def _find_used_source_rse(self, session, logger):
 
     @staticmethod
-    def _dst_file_set_and_file_corrupted(request, dst_file):
+    def _dst_file_set_and_file_corrupted(request: dict, dst_file: dict):
```
return type: `bool`

rdimaio commented on lib/rucio/transfertool/fts3.py:
```diff
@@ -572,7 +574,7 @@ def _dst_file_set_and_file_corrupted(request, dst_file):
 
     @staticmethod
-    def _dst_file_set_and_file_correct(request, dst_file):
+    def _dst_file_set_and_file_correct(request: dict, dst_file: dict):
```
return type: `bool`

rdimaio commented on lib/rucio/transfertool/fts3.py:
```diff
@@ -629,7 +631,7 @@ def __init__(self, external_host, request_id, fts_message):
         self._dst_url = fts_message.get('dst_url', None)
 
-    def initialize(self, session, logger=logging.log):
+    def initialize(self, session: "Session", logger: Callable = logging.log):
```
return type: `None`

rdimaio commented on lib/rucio/transfertool/fts3.py:
```diff
@@ -1454,7 +1464,7 @@ def __bulk_query_responses(self, jobs_response, requests_by_eid):
         return responses
 
-    def __query_details(self, transfer_id):
+    def __query_details(self, transfer_id: str):
```
return type: `Optional[dict[str, Any]]`

rdimaio commented on lib/rucio/transfertool/fts3_plugins.py:
```diff
@@ -0,0 +1,152 @@
+        Register a fts3 transfer plugin
+
+        :param name: name to register under
+        :type name: str
```
I think the `type` notes in the docstring might be unnecessary given that the types are already mentioned in the function arguments

rdimaio commented on lib/rucio/transfertool/fts3_plugins.py:
```diff
@@ -0,0 +1,152 @@
+        except NoSectionError:
+            self.prority_table = {}
+
+    def _activity_hints(self, activity_kwargs: dict[str, str], default_prority: str = '20') -> dict[str, dict]:
```
return type `dict[str, dict[str, str]]` would be more accurate

rdimaio commented on lib/rucio/transfertool/fts3_plugins.py:
```diff
@@ -0,0 +1,152 @@
+        """
+        return {"collocation_hints": collocation_func(*hints)}
+
+    def _test_collocation(self, *hint: dict) -> dict:
```
a more accurate return type: `dict[str, str]`

rdimaio commented on lib/rucio/transfertool/fts3_plugins.py:
```diff
@@ -0,0 +1,152 @@
+    def _test_collocation(self, *hint: dict) -> dict:
+        return {"0": "", "1": "", "2": "", "3": ""}
+
+    def _default(self, *hints: dict) -> dict:
```
if this is meant to be implemented in the child classes, it might be worth using the [`@abstractmethod` decorator](https://docs.python.org/3/library/abc.html#abc.abstractmethod)

rdimaio commented on lib/rucio/transfertool/fts3_plugins.py:
```diff
@@ -0,0 +1,152 @@
+                f"Request too large, decrease to less than {self.transfer_limit}", e
+            )
+
+    def hints(self, hint_kwargs: dict) -> dict:
```
return type `dict[str, Any]` would be more accurate

rdimaio commented on tests/test_transfer_plugins.py:
```diff
@@ -0,0 +1,271 @@
+
+    job_params = fts3_tool._file_from_transfer(transfer_params, job_params)
+
+    # Exstract the hints
```
typo: "Exstract" -> "Extract"

voetberg replied:
😅

rdimaio commented on tests/test_transfer_plugins.py:
```diff
@@ -0,0 +1,271 @@
+    # Get the job params used for each transfer
+    job_params = fts3_tool._file_from_transfer(transfer_path[0], job_params)
+
+    # Exstract the hints
```
Typo: "Exstract" -> "Extract"

voetberg committed:
Transfers: Plugins for interacting with fts3 'archive_metdata' field #6398

Changes:
	* Introduce fts3_plugin class, a subclass of PolicyPlugins
	* Include collocation wrapper, prority placement algorithm for
	  activity based collocation/placement
	* Typing for fts3_transfertool in part

bari12 commented:
The multi_vo tests seem to be failing now. Please investigate

voetberg commented:
> The multi_vo tests seem to be failing now. Please investigate

According to the logs they look like they're failing before they even get to the tests (Looks like the docker-compose hits an error with getting the activemq image.)

As for the integration test - Seems unrelated? [The version on fork passed without problem. ](https://github.com/voetberg/rucio/actions/runs/8098623093/job/22132464803)

rdimaio commented:
Looking into the autotest failures (these two for reference, [1](https://github.com/rucio/rucio/actions/runs/8098624010/job/22185879591?pr=6445) and [2](https://github.com/voetberg/rucio/actions/runs/8143245286/job/22254672967))

1. `test.sh` starts: `docker exec c37ff6ff8585cb8d_rucio_1 ./tools/test/test.sh`
2. gets to `Waiting for httpd`
3. Exits: `** Process '['docker', 'exec', 'c37ff6ff8585cb8d_rucio_1', './tools/test/test.sh']' exited with code 23 {'DIST': 'alma9', 'PYTHON': '3.9', 'IMAGE_IDENTIFIER': 'autotest', 'SUITE': 'multi_vo', 'RDBMS': 'postgres14', 'RUCIO_HOME': '/opt/rucio/etc/multi_vo/tst', 'IMAGE': 'ghcr.io/voetberg/rucio/rucio-autotest:alma9-python3.9'}`

For `multi_vo`, the `wait_for_httpd` line is here: https://github.com/rucio/rucio/blob/master/tools/test/test.sh#L115, followed by
```
httpd -k restart
tools/run_multi_vo_tests_docker.sh
```

`tools/run_multi_vo_tests_docker.sh` echoes `Clearing memcache` as initial message: https://github.com/rucio/rucio/blob/master/tools/run_multi_vo_tests_docker.sh#L39 

so it seems like the`test.sh` fails before entering `tools/run_multi_vo_tests_docker.sh`.

Here's an example of this `multi_vo` test for another PR: https://github.com/rucio/rucio/actions/runs/7755635332/job/21151415317#step:4:84 - here `tools/run_multi_vo_tests_docker.sh` is entered successfully

rdimaio commented:
There are only two commands running between `echo 'Waiting for httpd'` and `tools/run_multi_vo_tests_docker.sh`:

1. `curl --output /dev/null --silent --retry 15 --retry-all-errors --retry-delay 1 -k https://localhost/ping`
2. `httpd -k restart`

`httpd` has no exit code `23`: https://nightlies.apache.org/httpd/trunk/doxygen/group__HTTP__Status.html

`curl` has exit code `23`: https://curl.se/docs/manpage.html and https://everything.curl.dev/cmdline/exitcode
> Write error. Curl could not write data to a local filesystem or similar. curl receives data chunk by chunk from the network and it stores it like at (or writes it to stdout), one piece at a time. If that write action gets an error, this is the exit status.

So the issue is occurring here: https://github.com/rucio/rucio/blob/master/tools/test/test.sh#L34

voetberg commented:
That makes sense, but I suppose the question is - why is this only an issue on this PR that doesn't (or shouldn't rather), modify that at all? I'll admit I can't think of a reason, especially when it fails consistently on both my fork and this PR running the canonical version of the tests.

rdimaio commented:
I pulled your branch and pushed it to my upstream fork, just to see if maybe it was some weird issue with the fork itself - I get the same failure on my fork: https://github.com/rdimaio/rucio/actions/runs/8144295087/job/22258118979

I made a commit to remove `--output /dev/null --silent` (commit for reference: https://github.com/rdimaio/rucio/commit/79e109cb716f943607be853ebd56c2c92153701e), here's the test run: https://github.com/rdimaio/rucio/actions/runs/8144460703/job/22258608666 - it works here, so the issue is definitely with outputting the curl output to `/dev/null`. I have no idea why this is only affecting this branch, but that would be a quick fix to resolve it.

Side note - the test logic involved here (e.g. the `curl` command) is related to the token infrastructure changes (https://github.com/rucio/rucio/blame/master/tools/test/test.sh#L115), so it might be a good idea to get someone that worked on that to share their opinion on this (@bari12 tagging you in case you know anyone other than Radu who could have some insight here)

voetberg commented:
Thanks @rdimaio ! I was about to put in a PR for the same. Glad you were able to replicate it on your fork - I was loosing my mind. 

The only thing I can think that causes is maybe running out of disk on github's ends, as the mutli-vo tests are the last things run and is a larger commit? Might be nonsense but besides for a permission error that's one of the few things that can cause that `23` error code.

voetberg committed:
Tests: removed curl writing to dev/null
